pyomo/solvers/plugins/solvers/GAMS.py

```diff
@@ -0,0 +1,792 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from __future__ import print_function
+from six import StringIO
+from pyutilib.misc import PauseGC
+from pyomo.core.base import (
+    SymbolMap, AlphaNumericTextLabeler, NumericLabeler,
+    Block, Constraint, Expression, Objective, Var, Set, RangeSet, Param,
+    value, minimize, Suffix)
+from pyomo.core.base.component import ComponentData
+import os, sys, subprocess, pipes, math, logging
+
+import pyomo.util.plugin
+from pyomo.opt.base import IOptSolver
+import pyutilib.services
+
+from pyutilib.misc import Options
+
+
+logger = logging.getLogger('pyomo.solvers')
+
+pyutilib.services.register_executable(name="gams")
+ 
+class GAMSSolver(pyomo.util.plugin.Plugin):
+    """A generic interface to GAMS solvers"""
+    pyomo.util.plugin.implements(IOptSolver)
+    pyomo.util.plugin.alias('gams', doc='The GAMS modeling language')
+
+    options = Options()
+
+    def available(self, exception_flag=True):

```

Comment:
In theory you can use the same file, however the two solver interfaces should be separate classes so that they can be registered separately with the `SolverFactory`.  One option is to make both interfaces "private" and create a third solver that redirects to the correct solver based on the `solver_io` option (see `pyomo/solvers/plugins/solvers/CPLEX.py` for an example).

That said, I don't think that we should spend too much time worrying about it: we are probably going to do a massive overhaul of the solver class hierarchy "soon" to fix some of this weirdness (which would make much of this discussion moot).

@gseastream: `register_executable` is a PyUtilib method that tells an "excutable factory" that you would like it to try and find the listed executable somewhere on the PATH.  We use it because it standardizes all the neurotic platform-specific weirdness.  `registered_executable` will return an object that among other things has the full path to the located executable, or `None` if PyUtilib couldn't find the executable.